src/NuGet.Core/NuGet.Packaging/Rules/ReferencesInNuspecMatchRefAssetsRule.cs

```diff
@@ -0,0 +1,137 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Text;
+using System.Threading.Tasks;
+using System.Xml.Linq;
+using NuGet.Client;
+using NuGet.Common;
+using NuGet.ContentModel;
+using NuGet.Frameworks;
+using NuGet.Packaging.Core;
+using NuGet.RuntimeModel;
+
+namespace NuGet.Packaging.Rules
+{
+    class ReferencesInNuspecMatchRefAssetsRule : IPackageRule
+    {
+        private string _addToRefFormat = AnalysisResources.ReferencesInNuspecAndRefFilesDontMatchWarningAddToRefListItemFormat;
+        private string _addToNuspecFormat = AnalysisResources.ReferencesInNuspecAndRefFilesDontMatchWarningAddToNuspecListItemFormat;
+        public string MessageFormat => AnalysisResources.ReferencesInNuspecAndRefFilesDontMatchWarning;
+
+        public IEnumerable<PackagingLogMessage> Validate(PackageArchiveReader builder)
+        {
+            var refFiles = builder.GetFiles().Where(t => t.StartsWith(PackagingConstants.Folders.Ref));
+            var nuspecReferences = GetReferencesFromNuspec(builder.GetNuspec());
+            var missingItems = Compare(nuspecReferences, refFiles);
+            return GenerateWarnings(missingItems);
+        }
+
+        internal IDictionary<string, string[]> GetReferencesFromNuspec(Stream nuspecStream)
+        {
+            var nuspecReferences = new Dictionary<string, string[]>();
+            var packageNuspec = new NuspecReader(nuspecStream);
+            var nuspec = packageNuspec.Xml;
+            if (nuspec != null)
+            {
+                XNamespace name = nuspec.Root.Name.Namespace;
+                var keys = nuspec.Descendants(XName.Get("{" + name.NamespaceName + "}references")).Elements().Attributes("targetFramework");
+                var shortName = keys.Select(t => NuGetFramework.Parse(t.Value).GetShortFolderName());
+                nuspecReferences = shortName.Zip(keys.Select(t => t.Parent.Elements().Attributes("file").Select(m => m.Value)), (tfm, refFiles) => new { tfm, refFiles = refFiles.ToArray() })
```

Comment:
easy to understand code is more important than short code. I feel like this would be easier to understand as a `foreach` loop, rather than `Zip` and `ToDictionary`.

Reply:
alternatively
```csharp
keys.ToDictionary(
  k => NuGetFramework.parse(k.Value).GetShortFolderName(),
  k => k.Parent.Elements().Attributes("file").Select(f => f.Value));

```